File: mobile/ios/Shell/Shared/ModeStore.swift
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 dlf-dds contributors.
//
// ModeStore persists the operator-selected OperatingMode to the App Group
// UserDefaults so the NEPacketTunnelProvider (running in a separate
// extension process from the main app) reads the same value the UI wrote.
// Mirrors BundleStore's "main-app writes, extension reads" pattern.

import Foundation

struct ModeStore {
    private static let key = "io.dlf-dds.goat-client.operating-mode"

    /// Read the persisted mode. Returns `OperatingMode.default` when nothing
    /// has been written yet (fresh install) or when the App Group defaults
    /// are unavailable (Simulator misconfigured entitlement — same surface
    /// as BundleStore's noContainer case).
    static func read() -> OperatingMode {
        guard let defaults = defaults(),
              let raw = defaults.string(forKey: key),
              let mode = OperatingMode(rawValue: raw) else {
            return .default
        }
        return mode
    }

    /// Persist `mode` for the next tunnel start. Cheap and idempotent.
    /// Returns false only if the App Group defaults are unavailable.
    @discardableResult
    static func write(_ mode: OperatingMode) -> Bool {
        guard let defaults = defaults() else { return false }
        defaults.set(mode.rawValue, forKey: key)
        return true
    }

    /// Reset to the default mode. Used by the test harness and by `Clear
    /// bundle` (so a freshly imported bundle in a new posture doesn't keep
    /// stale mode state from the previous import).
    static func reset() {
        defaults()?.removeObject(forKey: key)
    }

    private static func defaults() -> UserDefaults? {
        UserDefaults(suiteName: AppGroup.identifier)
    }
}

File: mobile/android/Shell/app/src/main/java/io/dlf_dds/goat_client/ModeStore.kt
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 dlf-dds contributors.
package io.dlf_dds.goat_client

import android.content.Context
import android.content.SharedPreferences

/**
 * Persists the operator-selected [OperatingMode] to an app-private
 * SharedPreferences file the [MainActivity] writes and [GoatVpnService]
 * reads. Process-shared because both Activity and Service live in the
 * default `:app` process — no need for a per-user MODE_MULTI_PROCESS
 * shim. Mirrors the iOS shell's ModeStore (App Group UserDefaults).
 */
object ModeStore {
    private const val PREFS_NAME = "io.dlf_dds.goat_client.prefs"
    private const val KEY_MODE = "operating-mode"

    fun read(ctx: Context): OperatingMode {
        val raw = prefs(ctx).getString(KEY_MODE, null)
        return OperatingMode.fromRaw(raw) ?: OperatingMode.DEFAULT
    }

    fun write(ctx: Context, mode: OperatingMode) {
        prefs(ctx).edit().putString(KEY_MODE, mode.raw).apply()
    }

    /** Reset to the default mode (used by `Clear bundle`). */
    fun reset(ctx: Context) {
        prefs(ctx).edit().remove(KEY_MODE).apply()
    }

    private fun prefs(ctx: Context): SharedPreferences =
        ctx.applicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
}
